tests/Unit/Lti/Factory/Lti1p3RequestFactoryTest.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+/**
+ *  This program is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU General Public License
+ *  as published by the Free Software Foundation; under version 2
+ *  of the License (non-upgradable).
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ *  Copyright (c) 2020 (original work) Open Assessment Technologies S.A.
+ */
+
+declare(strict_types=1);
+
+namespace OAT\SimpleRoster\Tests\Unit\Lti\Factory;
+
+use Carbon\Carbon;
+use OAT\Library\Lti1p3Core\Message\Launch\Builder\LtiResourceLinkLaunchRequestBuilder;
+use OAT\Library\Lti1p3Core\Message\LtiMessageInterface;
+use OAT\Library\Lti1p3Core\Registration\RegistrationInterface;
+use OAT\Library\Lti1p3Core\Registration\RegistrationRepositoryInterface;
+use OAT\SimpleRoster\Entity\Assignment;
+use OAT\SimpleRoster\Entity\LineItem;
+use OAT\SimpleRoster\Entity\User;
+use OAT\SimpleRoster\Lti\Exception\RegistrationNotFoundException;
+use OAT\SimpleRoster\Lti\Factory\Lti1p3RequestFactory;
+use OAT\SimpleRoster\Lti\Request\LtiRequest;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class Lti1p3RequestFactoryTest extends TestCase
+{
+    /** @var Lti1p3RequestFactory */
+    private $subject;
+
+    /** @var LtiResourceLinkLaunchRequestBuilder|MockObject */
+    private $builder;
+
+    /** @var RegistrationRepositoryInterface|MockObject */
+    private $repository;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->builder = $this->createMock(LtiResourceLinkLaunchRequestBuilder::class);
+        $this->repository = $this->createMock(RegistrationRepositoryInterface::class);
+
+        $this->subject = new Lti1p3RequestFactory($this->repository, $this->builder, 'registrationId');
+    }
+
+    /**
+     * @dataProvider provideValidLineItemAssignmentCombinations
+     */
+    public function testItReturnsAssignmentLtiRequest(
+        int $maxAttempts,
+        int $attemptsCount,
+        string $assignmentStatus
+    ): void {
+        Carbon::setTestNow(Carbon::createFromDate(2020, 1, 1));
+
+        $assignment = $this->createAssignmentMock($maxAttempts, $attemptsCount, $assignmentStatus);
+
+        $this->createRegistrationMock();
+
+        $message = $this->createMock(LtiMessageInterface::class);
+        $message
+            ->expects($this->once())
```

Comment:
`self::once()` (please fix it everywhere)

Reply:
Can you tell me why? And  should do it to everything? `self::createMock()`, `self::assertSomething()`?

Reply:
```
/**
     * Returns a matcher that matches when the method is executed exactly once.
     */
    public static function once(): InvokedCountMatcher
    {
        return new InvokedCountMatcher(1);
    }
```